Update GroundHeatExchangerHorizontalTrench: wrap Site:GroundTemperature:Undisturbed:XXX objects

joseph-robertson opened the pull request:
Pull request overview
---------------------

Companion PR:
* https://github.com/NREL/OpenStudio/pull/4717

Link to relevant GitHub Issue(s) if appropriate: https://github.com/NREL/OpenStudio/pull/4717

Link to the **Linux.deb** installer to use for CI Testing. If not set, it will default to latest official release.
[OpenStudio Installer]: http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-4717/OpenStudio-3.5.2-alpha%2Be00b5b37e4-Ubuntu-18.04.deb

This Pull Request is concerning:

 - [x] **Case 1 - `NewTest`:** a new test for a new model API class,

----------------------------------------------------------------------------------------------------------

### Case 1: New test for a new model API class

Please include which class(es) you are adding a test to specifically test for.
Include a link to the OpenStudio Pull Request in which you are adding the new classes, or the class itself if already on develop.
<!---
> eg:
>
> This pull request is in relation with the Pull Request [NREL/OpenStudio#3031](https://github.com/NREL/OpenStudio/pull/3031), and  will specifically test for the following classes:
> * `AirTerminalSingleDuctConstantVolumeFourPipeBeam`
> * `CoilCoolingFourPipeBeam`
> * `CoilHeatingFourPipeBeam`
> * Additionally it explicitly tests for the existing class [TableMultiVariableLookUp](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/TableMultiVariableLookup.hpp)
-->

#### Work Checklist

The following has been checked to ensure compliance with the guidelines:

 - [ ] Tests pass either:
     - [ ] with official OpenStudio release (include version):
         - [ ] A matching OSM test has been added from the successful run of the Ruby one with the official OpenStudio release
         - [ ] The label `AddedOSM` has been added to this PR
         - [ ] All new `out.osw` have been committed

     - [ ] with current develop (incude SHA):
         - [ ] The label `PendingOSM` has been added to this PR
         - [ ] A matching OSM test has not yet been added because the official release is pending, but `model_tests.rb` has a TODO.
            ```ruby
            def test_airterminal_cooledbeam_rb
              result = sim_test('airterminal_cooledbeam.rb')
            end

            # TODO: To be added in the next official release after: 2.5.0
            # def test_airterminal_fourpipebeam_osm
            #   result = sim_test('airterminal_fourpipebeam.osm')
            # end
            ```
        - [ ] No `out.osw` have been committed as they need to be run with an official OpenStudio version


 - [ ] **Ruby test is stable**: when run multiple times on the same machine, it produces the same total site kBTU.
    Please paste the heatmap png generated after running the following commands:
     - [ ] I ensured that I assign systems/loads/etc in a repeatable manner (eg: if I assign stuff to thermalZones, I do `model.getThermalZones.sort_by{|z| z.name.to_s}.each do ...` so I am sure I put the same ZoneHVAC systems to the same zones regardless of their order)
     - [ ] I tested stability using `process_results.py` (see `python process_results.py --help` for usage).
    Please paste the text output or heatmap png generated after running the following commands:
        ```bash
        # Clean up all custom-tagged OSWs
        python process_results.py test-stability clean
        # Run your test 5 times in a row. Replace `testname_rb` (eg `airterminal_fourpipebeam_rb`)
        python process_results.py test-stability run -n testname_rb
        # Check that they all passed
        python process_results.py test-status --tagged
        # Check site kBTU differences
        python process_results.py heatmap --tagged

        ```

 - [ ] **Object has been added to `autosize_hvac.rb` to ensure the autosizedXXX values methods do work**

----------------------------------------------------------------------------------------------------------

### Review Checklist

 - [ ] Code style (indentation, variable names, strip trailing spaces)
 - [ ] Functional code review (it has to work!)
 - [ ] Matching OSM test has been added or `# TODO` added to `model_tests.rb`
 - [ ] Appropriate `out.osw` have been committed
 - [ ] Test is stable
 - [ ] Object is tested in `autosize_hvac` as appropriate
 - [ ] The appropriate labels have been added to this PR:
   - [ ] One of: `NewTest`, `TestFix`, `NewTestForExisting`, `Other`
   - [ ] If `NewTest`: add `PendingOSM` or `AddedOSM`

joseph-robertson committed:
Stub new ghx horizontal trench test file.

joseph-robertson committed:
Update model_tests rb.

jmarrec commented on model_tests.rb:
```diff
@@ -527,6 +527,15 @@ def test_generator_windturbine_osm
+    result = sim_test('ghx_horizontal_trench.rb')
+  end
+
+  # TODO: To be added in the next official release after: 3.5.0
```
```suggestion
  # TODO: To be added in the next official release after: 3.5.1
```

jmarrec reviewed (approved):
Code looks fine, model has minimal number of zones. Just need to demonstrate that it runs succesfully using the most recent installer form https://github.com/NREL/OpenStudio/pull/4717 and I'm good with it

joseph-robertson committed:
Merge branch 'develop' into ghx-horiz-trench

joseph-robertson committed:
Update comment.

joseph-robertson commented:
I'm running into errors.

Adding both site ground temperature undisturbed kusuda _and_ xing, I get:
```
   ** Severe  ** PlantPipingSystems::CreatePartitionRegionList: Invalid partition location in domain.
   **   ~~~   ** Occurs during mesh development for domain=HorizontalTrenchDomain   1
   **   ~~~   ** A pipe or basement is located outside of the domain extents.
```

Adding just one, I get:
```
   ** Severe  ** Plant temperatures are getting far too hot, check controls and relative loads and capacities
```

joseph-robertson committed:
Explicitly set all site ground temperature properties.

jmarrec commented on model_tests.rb:
```diff
@@ -529,6 +529,15 @@ def test_generator_windturbine_osm
+    result = sim_test('ghx_horizontal_trench.rb')
+  end
+
+  # TODO: To be added in the next official release after: 3.5.1
+  # def test_ghx_horizontal_trench_osm
+  #  result = sim_test('ghx_horizontal_trench.osm')
+  # end
```
I think for the Kusuda one we actually want to make a test that runs on 3.5.1 (at least there) and commit that OSM. So we can test your VT and make sure results didn't change too

jmarrec committed:
NOTE: IT HAS SWITCHES, REMOVE. Modify test to use a LoadProfile:Plant instead.... make it pass with 3.5.1.

jmarrec commented on model/simulationtests/ghx_horizontal_trench.rb:
```diff
@@ -0,0 +1,170 @@
+
+model = BaselineModel.new
+
+# No zones, just a LoadProfile:Plant
+# This is unfortunate but the only E+ example file that uses it is PlantHorizontalGroundHX.idf
+# and it has just a LoadProfile:Plant, and nothing autosized
+# We couldn't get this file to not throw a Plant run away temperature so we're
+# matching the E+ test instead
```
@joseph-robertson

jmarrec commented on model/simulationtests/ghx_horizontal_trench.rb:
```diff
@@ -0,0 +1,170 @@
+# add design days to the model (Chicago)
+model.add_design_days
+
+USE_KUSUDA = true
+USE_XING = false
+USE_PIPE_INDOOR = false
+
+raise "Cannot use XING on 3.5.1 and below" if USE_XING and Gem::Version.new(OpenStudio.openStudioVersion) <= Gem::Version.new('3.5.1')
```
Ok so for now I have left these switches, until we are satisfied with what we're doing and ready to commit to several files.

I suppose eventually after we're done testing we need to make two ruby tests: A kusuda one, and a Xing one.

The kusuda should be run at 3.5.1, and the osm saved, and results commited.

the Xing requires the next official version so only the ruby one should be added.

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.7p221 (2022-11-24 revision 168ec2b1e5) [x86_64-linux], Rubocop 0.81.0)

joseph-robertson committed:
Test both kusuda and xing.

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.7p221 (2022-11-24 revision 168ec2b1e5) [x86_64-linux], Rubocop 0.81.0)

jmarrec committed:
Add the Kusuda OSM test at 3.5.0. It was failing before. Run kusuda rb + osm at 3.5.0 and 3.5.1 official

Commit that fixes it in 3.5.0: https://github.com/NREL/OpenStudio/commit/65bcf102dbde3fe69d83d5ef61ed09694588e925

jmarrec commented:
I commited the Kusuda OSM from 3.5.0 (it was failing before, cf https://github.com/NREL/OpenStudio/commit/65bcf102dbde3fe69d83d5ef61ed09694588e925)

Tests pass:

![image](https://user-images.githubusercontent.com/5479063/219046677-cf362146-13d9-4f4b-adf5-c46dd173a0ee.png)


VT is functioning properly too! The IDFs at 3.5.2-alpha between the kusuda_rb and kusuda_osm are identical.

![image](https://user-images.githubusercontent.com/5479063/219046829-5b0c8f27-a0f0-489d-886c-433f8f4e815b.png)


I was a bit confused by the fact that both the kusuda and xing produce the same results, but I can confirm the IDFs are different:

```shell
$ diff -u testruns/ghx_horizontal_trench_kusuda.rb/run/in.idf testruns/ghx_horizontal_trench_xing.rb/run/in.idf 
```

```diff
--- testruns/ghx_horizontal_trench_kusuda.rb/run/in.idf	2023-02-15 14:47:40.883324729 +0100
+++ testruns/ghx_horizontal_trench_xing.rb/run/in.idf	2023-02-15 14:47:41.363330865 +0100
@@ -547,14 +547,16 @@
   Node 3,                                 !- Component Inlet Node Name 1
   Node 8;                                 !- Component Outlet Node Name 1
 
-Site:GroundTemperature:Undisturbed:KusudaAchenbach,
-  Site Ground Temperature Undisturbed Kusuda Achenbach 1, !- Name
+Site:GroundTemperature:Undisturbed:Xing,
+  Site Ground Temperature Undisturbed Xing 1, !- Name
   1.08,                                   !- Soil Thermal Conductivity {W/m-K}
   962,                                    !- Soil Density {kg/m3}
   2576,                                   !- Soil Specific Heat {J/kg-K}
-  15.5,                                   !- Average Soil Surface Temperature {C}
-  12.8,                                   !- Average Amplitude of Surface Temperature {deltaC}
-  17.3;                                   !- Phase Shift of Minimum Surface Temperature {days}
+  11.1,                                   !- Average Soil Surface Temperature {C}
+  13.4,                                   !- Soil Surface Temperature Amplitude 1 {deltaC}
+  0.7,                                    !- Soil Surface Temperature Amplitude 2 {deltaC}
+  25,                                     !- Phase Shift of Temperature Amplitude 1 {days}
+  30;                                     !- Phase Shift of Temperature Amplitude 2 {days}
 
 GroundHeatExchanger:HorizontalTrench,
   Ground Heat Exchanger Horizontal Trench 1, !- Name
@@ -575,8 +577,8 @@
   2405,                                   !- Pipe Specific Heat {J/kg-K}
   30,                                     !- Soil Moisture Content Percent {percent}
   50,                                     !- Soil Moisture Content Percent at Saturation {percent}
-  Site:GroundTemperature:Undisturbed:KusudaAchenbach, !- Undisturbed Ground Temperature Model Type
-  Site Ground Temperature Undisturbed Kusuda Achenbach 1, !- Undisturbed Ground Temperature Model Name
+  Site:GroundTemperature:Undisturbed:Xing, !- Undisturbed Ground Temperature Model Type
+  Site Ground Temperature Undisturbed Xing 1, !- Undisturbed Ground Temperature Model Name
   0.408;                                  !- Evapotranspiration Ground Cover Parameter
```

![image](https://user-images.githubusercontent.com/5479063/219047316-4f6daeb0-a5f2-410e-b798-e18d4e7d9f31.png)
